modules/host/contractmanager/sectorupdate.go

```diff
@@ -411,10 +417,15 @@ func (cm *ContractManager) AddSectorBatch(sectorRoots []crypto.Hash) error {
 
 	// Add each sector in a separate goroutine.
 	var wg sync.WaitGroup
+	semaphore := make(chan struct{}, maxParallelThreads)
 	for _, root := range sectorRoots {
 		wg.Add(1)
 		go func(root crypto.Hash) {
 			defer wg.Done()
+			semaphore <- struct{}{}
```

Comment:
Need to request the semaphore before you call `go func`, otherwise you're still going to end up with thousands of goroutines, they just won't be doing anything.